Sources/HummingbirdRouter/RouteController.swift

```diff
@@ -0,0 +1,106 @@
+/// type for the controller's `Context`.
+///
+///     struct UsersController {
+///         typealias Context = BasicRouterRequestContext
+///
+///         var body: some RouterMiddleware<Context> {
+///             RouteGroup("/users") {
+///                 ...
+///             }
+///         }
+///     }
+///
+/// Assemble the controller's body by combining one or more components together to create
+/// a route. By nesting controllers within each other, you can compose large and complex
+/// routes from smaller more mangeable components.
+public protocol RouteController where Context == Body.Context {
+    associatedtype Context: RouterRequestContext
+    associatedtype Body: RouterMiddleware
+    @RouteBuilder<Context> var body: Body { get }
+}
+
+// MARK: - MiddlewareFixedTypeBuilder + RouteController
+
+/// Middleware stack result builder
+///
+/// Generates a middleware stack from the elements inside the result builder. The input,
+/// context and output types passed through the middleware stack are fixed and cannot be changed.
+extension MiddlewareFixedTypeBuilder {
+    public static func buildExpression<C0: RouteController>(_ c0: C0) -> C0.Body where C0.Body.Input == Input, C0.Body.Output == Output, C0.Body.Context == Context {
+        return c0.body
+    }
+
+    public static func buildBlock<C0: RouteController>(_ c0: C0) -> C0.Body {
+        return c0.body
+    }
+
+    public static func buildPartialBlock<C0: RouteController>(first: C0) -> C0.Body {
+        first.body
+    }
+
+    public static func buildPartialBlock<C0: RouteController, C1: RouteController>(
+        accumulated c0: C0,
+        next c1: C1
+    ) -> _Middleware2<C0.Body, C1.Body> where C0.Body.Input == C1.Body.Input, C0.Body.Output == C1.Body.Output, C0.Body.Context == C1.Body.Context {
+        _Middleware2(c0.body, c1.body)
+    }
+
+    public static func buildPartialBlock<C0: RouteController, M0: MiddlewareProtocol>(
+        accumulated c0: C0,
+        next m0: M0
+    ) -> _Middleware2<C0.Body, M0> where C0.Body.Input == M0.Input, C0.Body.Output == M0.Output, C0.Body.Context == M0.Context {
+        _Middleware2(c0.body, m0)
+    }
+
+    public static func buildPartialBlock<M0: MiddlewareProtocol, C0: RouteController>(
+        accumulated m0: M0,
+        next c0: C0
+    ) -> _Middleware2<M0, C0.Body> where M0.Input == C0.Body.Input, M0.Output == C0.Body.Output, M0.Context == C0.Body.Context {
+        _Middleware2(m0, c0.body)
+    }
+}
+
+// MARK: - RouteBuilder + RouteController
+
+extension RouteBuilder {
+    public static func buildExpression<C0: RouteController>(_ c0: C0) -> C0.Body where C0.Body.Input == Request, C0.Body.Output == Response, C0.Body.Context == Context {
+        c0.body
+    }
+
+    public static func buildPartialBlock<C0: RouteController>(first: C0) -> C0.Body {
+        first.body
+    }
+
+    public static func buildPartialBlock<C0: RouteController, C1: RouteController>(
```

Comment:
The accumulated middleware should be a MiddlewareProtocol.

Reply:
Although I'm not sure we need to support this for the RouteBuilder result builder. This is only ever used in the situation where the endpoint is already selected eg

```swift
Get("test") {
    MyMiddleware()
    Handle { request, context in
        ...
    }
```